vacuumlabs/trezor-firmware
# Allow passing token account info into Trezor

Adds the option to pass `SolanaTxTokenAccountInfo` to Trezor so Trezor could verifiably display an associated token account's base address in token transfer transactions.

Accompanying Trezor Suite PR - https://github.com/trezor/trezor-suite/pull/10164.

Given `base_address`, `token_program`, `token_mint` Trezor derives the associated token account and checks if it matches the `token_account` in the transfer instruction. If token info for the triplet `(token_program, token_mint, `token_account)` is provided then the derived `token_account` must match the provided one. Otherwise we ignore the token info i.e. allow the tx to be signed but the token address is displayed.

If a base address is displayed instead of a token address we add an `info` button under which the associated token account is shown.

_I didn't add tests, comments (in .proto), docs yet, just so I wouldn't need to modify them after getting feedback._

I also realised that the "predefined" transaction check wouldn't match a token transfer if sending from multiple accounts which Suite supports - I have a draft for a fix already, will refactor and push.

**How to test**
You can also test through Trezor Suite.

Create a additional info file with the following content:
```
{
  "token_accounts_infos": [
    {
      "base_address": "BkoECWJYM7w9qNZ6EGCoBtnkphWau6nZMPbPqvm4eYF4",
      "token_program": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
      "token_mint": "GHArwcWCuk9WkUG4XKUbt935rKfmBmywbEWyFxdH3mou",
      "token_account": "92YgwqTtTWB7qY92JT6mbL2WCmhAs7LPZL4jLcizNfwx"
    }
  ]
}
```
And try signing the following transaction:

```
poetry run trezorctl solana sign-tx -n "m/44'/501'/0'/0'" -a additional_info.json  0100020500d1699dcb1811b50bb0055f13044463128242e37a463b52f6c97a1f6eef88ad5a20a57133eddc36177a6ac22923337e1c08107912f000722aaecb8ca3522ff57742e821a8ee6055c7e288aae92b85a8befa334096ee754b16c2d4f8ba2e4a83e303bf2ce35eddf29e99210c22a5203541ea2e3313bfed187a500eaf1b7e356006ddf6e1d765a193d9cbe146ceeb79ac1cb485ed5f5b37913a8cf5857eff00a91aea57c9906a7cad656ff61b3893abda63f4b6b210c939855e7ab6e54049213d010404010302000a0c140000000000000002
```

The base address (`BkoECWJYM7w9qNZ6EGCoBtnkphWau6nZMPbPqvm4eYF4`) should be displayed instead of the token address as recipient.

If you leave out the `-a` switch from the sign tx command the token address should be displayed.

If you change the `base_address` in the info file tx singning should fail.

If you change any of the other values, the tx should go through but token address should be displayed as recipient - since Trezor didn't get proper info for base address derivation.
